packages/core/content-releases/server/src/services/release.ts

```diff
@@ -500,198 +574,165 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => {
-        const releaseWithPopulatedActionEntries = (await strapi.entityService.findOne(
-          RELEASE_MODEL_UID,
-          releaseId,
-          {
-            populate: {
-              actions: {
-                populate: {
-                  entry: {
-                    fields: ['id'],
-                  },
-                },
-              },
-            },
-          }
-        )) as unknown as Release;
+      const trx = await strapi.db.transaction();
 
-        if (!releaseWithPopulatedActionEntries) {
+      try {
+        const lockedRelease = (await strapi.db
+          ?.queryBuilder(RELEASE_MODEL_UID)
+          .where({ id: releaseId })
+          .select(['id', 'releasedAt', 'status'])
+          .first()
+          .transacting(trx.get())
+          .forUpdate()
+          .execute()) as Pick<Release, 'id' | 'releasedAt' | 'status'> | undefined;
+
+        if (!lockedRelease) {
           throw new errors.NotFoundError(`No release found for id ${releaseId}`);
         }
 
-        if (releaseWithPopulatedActionEntries.releasedAt) {
+        if (lockedRelease.releasedAt) {
           throw new errors.ValidationError('Release already published');
         }
 
-        if (releaseWithPopulatedActionEntries.actions.length === 0) {
-          throw new errors.ValidationError('No entries to publish');
+        if (lockedRelease.status === 'failed') {
+          throw new errors.ValidationError(
+            'Release failed to publish before, change it to try again'
+          );
         }
 
-        /**
-         * We separate publish and unpublish actions, grouping them by contentType and extracting only their IDs. Then we can fetch more data for each entry
-         * We need to separate collectionTypes from singleTypes because findMany work as findOne for singleTypes and publishMany can't be used for singleTypes
-         */
-        const collectionTypeActions: {
-          [key: UID.ContentType]: {
-            entriestoPublishIds: ReleaseAction['entry']['id'][];
-            entriesToUnpublishIds: ReleaseAction['entry']['id'][];
-          };
-        } = {};
-        const singleTypeActions: {
-          uid: UID.ContentType;
-          id: ReleaseAction['entry']['id'];
-          action: ReleaseAction['type'];
-        }[] = [];
-        for (const action of releaseWithPopulatedActionEntries.actions) {
-          const contentTypeUid = action.contentType;
-
-          if (strapi.contentTypes[contentTypeUid].kind === 'collectionType') {
-            if (!collectionTypeActions[contentTypeUid]) {
-              collectionTypeActions[contentTypeUid] = {
-                entriestoPublishIds: [],
-                entriesToUnpublishIds: [],
-              };
-            }
+        // We need to pass the type because entityService.findOne is not returning the right type
+        const actions = await strapi.db.query(RELEASE_ACTION_MODEL_UID).findMany({
```

Comment:
Is there a reason we use the queryEngine here rather than the entityService used later on?

Reply:
Not a particular reason. Just that when rewriting this and considering that soon we are migrating this to v5 (where entityService is deprecated) and thought it was a good opportunity to change it 😅

Reply:
Releases, History, RW, ... and other internal features will not use content types but rather the lower DB Model implementation, so for those the document service will not be compatible.